src/bun.js/bindings/decodeURIComponentSIMD.cpp

```diff
@@ -0,0 +1,315 @@
+                cursor += 3;
+
+                // Get second byte
+                if (cursor + 2 >= end || *cursor != '%') {
+                    result.append(replacementChar);
+                    continue;
+                }
+
+                highNibble = hexToInt(cursor[1]);
+                lowNibble = hexToInt(cursor[2]);
+                if (highNibble > 15 || lowNibble > 15) {
+                    result.append(replacementChar);
+                    continue;
+                }
+                byte = (highNibble << 4) | lowNibble;
+                if ((byte & 0xC0) != 0x80) {
+                    result.append(replacementChar);
+                    continue;
+                }
+                value = (value << 6) | (byte & 0x3F);
+                cursor += 3;
+
+                // Check for overlong encoding
+                if (value < 0x80 || value > 0x7FF) {
+                    result.append(replacementChar);
+                    continue;
+                }
+
+                result.append(static_cast<UChar>(value));
+            } else if ((byte & 0xF0) == 0xE0) {
+                // 3-byte sequence
+                uint32_t value = byte & 0x0F;
+                cursor += 3;
+
+                // Get second byte
+                if (cursor + 2 >= end || *cursor != '%') {
+                    result.append(replacementChar);
+                    continue;
+                }
+                highNibble = hexToInt(cursor[1]);
+                lowNibble = hexToInt(cursor[2]);
+                if (highNibble > 15 || lowNibble > 15) {
+                    result.append(replacementChar);
+                    continue;
+                }
+                byte = (highNibble << 4) | lowNibble;
+                if ((byte & 0xC0) != 0x80) {
+                    result.append(replacementChar);
+                    continue;
+                }
+                value = (value << 6) | (byte & 0x3F);
+                cursor += 3;
+
+                // Get third byte
+                if (cursor + 2 >= end || *cursor != '%') {
+                    result.append(replacementChar);
+                    continue;
+                }
+                highNibble = hexToInt(cursor[1]);
+                lowNibble = hexToInt(cursor[2]);
+                if (highNibble > 15 || lowNibble > 15) {
+                    result.append(replacementChar);
+                    continue;
+                }
+                byte = (highNibble << 4) | lowNibble;
+                if ((byte & 0xC0) != 0x80) {
+                    result.append(replacementChar);
+                    continue;
+                }
+                value = (value << 6) | (byte & 0x3F);
+                cursor += 3;
+
+                // Check for overlong encoding and surrogate range
+                if (value < 0x800 || value > 0xFFFF || (value >= 0xD800 && value <= 0xDFFF) || // Surrogate range check
+                    (byte == 0xE0 && (value & 0x1F00) == 0)) // Overlong check for E0
```

Comment:
what is this check intended to be for? byte can never be 0xE0 because above it is checked if `(byte & 0xC0) != 0x80`